CompuCell3D/core/pyinterface/PlayerPythonNew/FieldWriterCML.cpp

```diff
@@ -0,0 +1,379 @@
+/**
+ * @file FieldWriterCML.cpp
+ * @author T.J. Sego, Ph.D.
+ * @brief Defines FieldWriterCML, a field writer for simulation-independent data throughput with VTK
+ * @date 2022-04-18
+ * 
+ */
+
+#include "FieldWriterCML.h"
+#include "FieldStorage.h"
+
+#include <CompuCell3D/Simulator.h>
+#include <CompuCell3D/Potts3D/Potts3D.h>
+#include <CompuCell3D/Field3D/Dim3D.h>
+
+#include <vtkDataArray.h>
+#include <vtkCharArray.h>
+#include <vtkDoubleArray.h>
+#include <vtkLongArray.h>
+#include <vtkPointData.h>
+
+
+#define FIELDWRITERCML_CHECKADD { if(!sim || !fsPtr || !latticeData) return false; }
+
+using namespace CompuCell3D;
+
+const std::string FieldWriterCML::CellTypeName = "CellType";
+const std::string FieldWriterCML::CellIdName = "CellId";
+const std::string FieldWriterCML::ClusterIdName = "ClusterId";
+
+FieldWriterCML::FieldWriterCML() : 
+    sim(NULL), 
+    fsPtr(NULL), 
+    latticeData(NULL)
+{}
+
+FieldWriterCML::~FieldWriterCML() {
+    this->clear();
+    if(sim != NULL) 
+        sim = NULL;
+    if(fsPtr != NULL) 
+        fsPtr = NULL;
+    if(latticeData != NULL) {
+        latticeData->Delete();
+    }
+}
+
+void FieldWriterCML::init(Simulator *_sim) {
+    sim = _sim;
+    latticeData = vtkStructuredPoints::New();
+    Dim3D fieldDim = sim->getPotts()->getCellFieldG()->getDim();
+    latticeData->SetDimensions(fieldDim.x, fieldDim.y, fieldDim.z);
+}
+
+void FieldWriterCML::clear() {
+    if(latticeData) 
+        for(auto &s : arrayNameVec) 
+            latticeData->GetPointData()->RemoveArray(s.c_str());
+    
+    arrayNameVec.clear();
+    arrayTypeVec.clear();
+}
+
+std::string FieldWriterCML::getFieldName(const unsigned int &i) const {
+    if(i >= numFields()) {
+        // todo: implement error on rebase
```

Comment:
I'm ok with approving it as is, but I'd prefer it if the errors were already implemented. It will save the users and us headaches